### ISSUE TITLE
Backwards incompatible change in patch version

### ISSUE THREAD

**Issue by prestwich** [paritytech/parity-scale-codec#520]
Updating from v3.6.4 -> v3.6.8 introduced a backwards incompatible change caught by ruint ci [here](https://github.com/recmo/uint/actions/runs/6151174710/job/16690707938?pr=310)

Please yank the patch and release as major

**Comment by ggwpez** [comment]
cc @pgherveou

**Comment by bkchr** [comment]
Yeah that was my fault :see_no_evil: https://github.com/paritytech/parity-scale-codec/pull/512#issuecomment-1710404546

Sorry!

**Comment by prestwich** [comment]
No worries, it's a pretty straightforward fix. Would help to know whether 3.6.8 will also be yanked, or if we should pin the dependency to `<3.6.8` in the next ruint release

**Comment by pgherveou** [comment]
@prestwich this ruint PR should fix it, not super familiar with the project, so reviews are welcome 🙏 
https://github.com/recmo/uint/pull/313

**Comment by bkchr** [comment]
Yeah maybe we should yank and release `4.0.0`. Or does anyone has better ideas?

**Comment by pgherveou** [comment]
Yanked it already to be safe, what about just fixing uint and then re-submitting?
Do you expect more projects that would start failing?

**Comment by pgherveou** [comment]
so, we should release v4.0.0 with the breaking change and ruint can update later on using the PR I submitted
Unless we want to rollback and go with the initial suggested commit of #508, which is not as nice but should not break downstream dependencies

**Comment by prestwich** [comment]
Yep, exactly. Let me know what ya'll end up deciding and I'll coordinate any ruint changes

**Comment by bkchr** [comment]
@pgherveou how much do you need `Compact` support for `MaxEncodedLen`? And could you maybe not use the `compact` attribute? This maybe already makes it work right now? Otherwise we can use your "less optimal" solution for now. There exists some issues that we should tackle before we go to `4.0`, we could also solve them and then bump.

**Comment by pgherveou** [comment]
We have this function here `read_sandbox_memory_as` that will read up to MaxEncodedLen bytes from memory to read and decode types from the contract memory. This would fail with a max_encode_len fn that returns a too small value for struct with compact fields.

It looks like we don't use Compact type now, and arbitrary contract storage do not rely on this method, so we should be safe for now, but I would like to read existing types that do define Compact field in the future.


https://github.com/paritytech/polkadot-sdk/blob/f7c95c5fc12d3abde866bc1ee106909f950e5427/substrate/frame/contracts/src/wasm/runtime.rs#L627-L638

**Comment by bkchr** [comment]
Without wanting to look to deep into this, but isn't `read_sandbox_memory_as` trusted code that is being used by `pallet-contracts` from the runtime side? Why do we need to use there the `MaxEncodedLen` at all?

**Comment by bkchr** [comment]
But why not just directly decode from the memory?

**Comment by pgherveou** [comment]
That's what we do, we decode the type from the contract's memory, we just need to make a bound check to make sure that we are not reading outside of the sandbox memory

**Comment by bkchr** [comment]
But at the point you are reading it, you are in trusted code? And you are declaring the type and not the contract. I really don't get why you need `MaxEncodedLen` for your use case.

**Comment by pgherveou** [comment]
```wat
(call $seal_terminate
	(i32.const 65536)  ;; Pointer to "account" address (out of bound).
)
```
Here reading `AccountId::max_encoded_len()` bytes from the account addr specified will trigger an outOfBound error.

**Comment by bkchr** [comment]
You can just change this code [here](https://github.com/paritytech/polkadot-sdk/blob/2ed66a0960367b2ea9fc23fa5ca59ee240b89db6/substrate/frame/contracts/src/wasm/runtime.rs#L609-L611) to this:

```
		let mut bound_checked = memory
			.get(ptr..)
			.ok_or_else(|| Error::<E::T>::OutOfBounds)?;
		let decoded = D::decode_with_depth_limit(MAX_DECODE_NESTING, &mut bound_checked)
``` 

`Decode` will return an error if there aren't enough bytes in the data to decode the type.
